### ISSUE TITLE
Faceted search wrong results

### ISSUE THREAD

**Issue by ghost** [PrestaShop/PrestaShop#15056]
Prestashop 1.7.6
PHP 7.2
Faceted module 3.2.0

![faceted](https://user-images.githubusercontent.com/8243273/62783364-8d5fe200-babc-11e9-904b-06d30cdf165a.png)

Module show Size 36 is available but after choosing product is only with size 39...

**Comment by 8ternity** [comment]
I was having issue with version 1.7.5.0 with Faceted Search. I was resolving it by purchase the module "[Amazzing filter Module](https://addons.prestashop.com/en/search-filters/18575-amazzing-filter.html)" and works really better ! Either if your native Prestahsop doesn't work. But you will need to disable or uninstall it after it. I know it's not related to resolving your issue, but after messing with Faceted Search on 1.7.5.0, 1.7.5.1 and 1.7.5.2 and having always an issue with it, i purchase a module which worth the money.

**Comment by khouloudbelguith** [comment]
Hi @PrestaShark,

In fact, in your case, the quantity of the General Product > 0
![image](https://user-images.githubusercontent.com/16067358/62787126-f004ae00-babb-11e9-9aa9-6bc105a0661c.png)
But, in your case, only one combination is out of stock which is `36`

When you make a filter with the `36` filter => yes, we have a product with the size, you are searching with the filer size & not the availability
but we have another filter which i availability => check if the product available or in stock.
I attached a screen record
https://drive.google.com/file/d/1VwHjCatMsdNcFm2KGTaYrv-zyWibkeLE/view
Thanks to check & feedback.

**Comment by ghost** [comment]
Hi.
Let me explain.

If someone search for exact size (no matter if its shoe or tshirt) then want to select wanted size and get matched products (in matter of UX need of use another filter like availability is highly unnecessary - searching within unavailble products is weird, and should be droped by the team working on UI/UX...)

So as customer i want to view shoes with size 42 then i want to click 42 and see all available shoes with this size. 

So i see (like in example) my size 36(1) click and get size 39 (as customer i can think that store is incompetent and leave it in future!). Customer is angry, admin is angry, both are wasting time.

It's obvious that the customer wants to look for only available products, so im treats my bug report very seriously - Sorry but it can't work like that anymore.

**Comment by ghost** [comment]
@marionf @colinegin @PierreRambaud please take care about it. 

If Faceted Search need more attention and time please consider to Crowdfund it and do it right. 

Thanks.

**Comment by ghost** [comment]
Selected size: 38.5 (1)
Selected availability: In stock/Available

And result as below :(

![Bez-nazwy-2](https://user-images.githubusercontent.com/8243273/62865514-0191c400-bd0f-11e9-9058-b07b069d6126.png)

**Comment by marionf** [comment]
@PierreRambaud This issue was supposed to be fixed: https://github.com/PrestaShop/PrestaShop/issues/10018

I can currently filter by a combination that is no longer available. This should not be possible if the combination is out of stock and if the order of out of stock products is not allowed

**Comment by Hlavtox** [comment]
@PrestaShark This is neverending story. I thought I would upgrade to 1.7 by 1.7.6 release, but it needs to wait. This software is just unusable in this condition.

- There is no properly working filtering. Dozens of combinations, bad product counts, no proper in-stock filter, page jumping to the top all the time.
- Translations are still not working after 2 years.
- Customers see delete icons next to permanent cart rules. Customers can see all vouchers in their account.
- Order note not working.
- Zero decimal places still not working.

So much invested money into modules and development and now we are stuck at old version, and PS's team argument is "look how many merchants use this". I would like to see which (successful) company uses this s***.

I don't understand why to convert more and more pages to symphony, and make more and more bugs which nobody fixes.

**Comment by ttoine** [comment]
FYI, this is quite possible to have a very successful website with PrestaShop 1.7.x

A few examples:

Spain:
- https://www.wonders.com
- https://www.cremas-caseras.es
- https://housepet.es/
- https://tienda-just-eat.es/
- https://www.zonatactica.es/
- https://charucashop.com/

Italy:
- https://www.soundtaste.eu/it/
- https://shop.imoving.it/
- https://www.rubinetteriashop.com/it/
- https://www.wrs.it/fr/
- https://www.maalbionline.com/it/

France:
- https://www.bobbies.com
- https://shop.magimix.fr/
- https://www.leslipfrancais.fr/
- https://www.papiertigre.fr
- https://www.recc-paris.com

Big companies like Decathlon are also using it for new websites, like Canada or Philippines.

When migrating a page to Symfony, most of the time, many bugs are fixed. But, you are right, sometimes, new bugs may be introduced. In this specific case, this is about the improvements done to a module, not about the core.

Thank a lot for your time to help us to fix this one.

**Comment by matks** [comment]
Also the bugs in Faceted Search are not related to Symfony, they are bugs introduced by a massive refactoring of the faceted search module codebase which definitely needed it. New version has been reworked for clearer and better implementation and has been covered by unit tests to catch as much bugs as possible. But the 0% bug target is an unreachable target.

About the disappointment you strongly state in `I would like to see which (successful) company uses this s***`:
I'll quote the wonderdull @dhh to answer this one: "Disappointment occurs when expectations don’t match reality. And our expectations for software quality are profoundly unrealistic. Thus, lots of people are continuously disappointed — even enraged — by software bugs. They shouldn’t be."
This quote comes from his blog post ["Software has bugs. This is normal."]
(https://m.signalvnoise.com/software-has-bugs-this-is-normal/) and is a great reading that I suggest you 😉 it will help you understand why there are bugs like this. Dhh is a wise man who founded the very successfull Basecamp product and the Ruby on Rails framework so I think we can agree he knows what he writes about, right 😉 ?

EDIT: I see @eternoendless just mentioned the very same blog post 😛 in the next answer

**Comment by eternoendless** [comment]
Another important note: PrestaShop is maintained by a very small team who does most of the heavy lifting on the project. We're not Microsoft or Uber, we don't have 6000 engineers working so that everything can move forward at the same time (although I'd very much like that!). That means things will take some time to be addressed. Oh and by the way, [bugs are normal](https://m.signalvnoise.com/software-has-bugs-this-is-normal/).

Here's the good news: this is an open source project, and people contribute by submitting fixes and improvements. This helps move things forward for interested parties. This means that if this problem is important for _you_, and if you consider it important to have it fixed soon, you have three choices:

- Either you fix it yourself (if you feel confident about it),
- Pay someone else to do it for you,
- ...or wait patiently that somebody else does it and shares it.

There's no way around it. The important part is _sharing_ so that the project moves forward so that _everyone_ can benefit.

**Comment by ghost** [comment]
Crowdfund is a keyword. Do it for crucial implementations. Do not hesitate to ask for it like Wikipedia do. Thirty Bees crowdfund elasticsearch implementation. Faceted search is most crucial right now guys...

I can't offer software that i can't rely on it :(
I love PS since 1.2.2.2. Make it great again.

**Comment by Hlavtox** [comment]
@ttoine @eternoendless @matks Sorry for the rude wording, but you really need to see the merchant's point of view. In every software there are bugs, even when I do a really small app, I fix myself multiple times. 

But there are bugs that are crucial and bugs that can wait. I know the bug classification, but you are making a software that people will use - general public. If something is not working in the backoffice, it can wait months and employees will just get used to it. But if there is a bug in the front end, you wont sell anything, customers wont understand it and will be pissed of and buy the stuff elsewhere.

**Read this examples and try to think about it:**

1. We don't have decimals in our currency. That makes Prestashop 1.7 completely (without modding it manually) unsuable for Czech Republic, Poland, Hungary etc.. To you think this is not something that should be implemented ASAP and not after 2 years?

2. How can I use it to offer goods to customers, if I can't change some nonsense wording, if the translation system still not work?

3. I am in the ecommerce world for 10 years and I spent 10 minutes deleting a product from my cart, that was a free gift. Then I realised it is a bug and it shows a trash icon, when it can't be deleted. What do you think that general stupid customer will do? Click it 3 times and say: "Fuck this store, I will order it elsewhere."

4. In any language other than English, customer order note is unreadable. This bug is open since Forge times. It is a matter of changing encoding/escaping somewhere in the code, I suppose changing like 3 lines of code. Just why?

5. Any customer in our store can go in to his Voucher screen in his account. He can see ALL VOUCHERS in the store, all vouchers with possible huge reductions that you "secretly" give to other customers. Do you understand how crucial mistake this is for the business? This is stuff that needs to be fixed in the next version, and not "it's on our roadmap".

6. I opened a ticket about layered navigation's page scrolling. Imagine you are a customer and you select a filter. What do you want to do next, see the products OR select another filter. So why THE FUCK the page scrolls intentionally to the top? Just WHY? And then @marionf tells me `No, it's not supposed to be fixed in 1.7.6.0. Please see this comment, the PR allow theme developer to choose to keep this scroll or not.` Show me one single person in the world that would want this behavior.

_____

PS: @matks The article is really nice, good read 👍

**Comment by matks** [comment]
> **Read this examples and try to think about it:**
> 
> 1. We don't have decimals in our currency. That makes Prestashop 1.7 completely (without modding it manually) unsuable for Czech Republic, Poland, Hungary etc.. To you think this is not something that should be implemented ASAP and not after 2 years?

If PrestaShop cannot be used in a major country like Poland it would be a critical issue indeed. But as far as I know it *is* being used right now, so I dont think we can say "PS 1.7 is unusable in Poland".

However I do not have a running Polish store, so I might be wrong on that. Consequently I will ask @kpodemski who is a renowned polish PrestaShop contributor whether this is a known issue and impacting issue. I think we can agree he's a good candidate to answer this question 😉 

> 2. How can I use it to offer goods to customers, if I can't change some nonsense wording, if the translation system still not work?

The translation system works. There are some bugs in it, some counter-intuitive behaviors and not everything is translatable. But it works. You can't say that a software with bugs in it "does not work", else Windows does not work, iOS does not work, Magento does not work, Linux does not work 😛 
If it was broken, people would, indeed, stop using PrestaShop and use something else and yet it is being used . Every time we break something in translation system, @kpodemski makes sure to remind us 😄 as this is a very important topic to him.

So there are tickets about known bugs in the prestashop translation system, they have been sorted in the backlog and they will be fixed accordingly. And these issues will get re-examined, like all issues, considering your feedback: every time we get feedback we re-evaluate which issue is critical and which is not.

> 3. I am in the ecommerce world for 10 years and I spent 10 minutes deleting a product from my cart, that was a free gift. Then I realised it is a bug and it shows a trash icon, when it can't be deleted. What do you think that general stupid customer will do? Click it 3 times and say: "Fuck this store, I will order it elsewhere."

I dont think we are talking about a critical bug here. This is a bug indeed, it has to be fixed, but it will be fixed:
- by us, when the other bugs which are more critical have been fixed
- or sooner if someone makes a contribution to the codebase that fixes this issue

Do you really want me to fix this issue rather than a issue that makes the price computation wrong ? 😛 Should I really spend time on something that is annoying the customer rather than making sure when customer orders product A at 10$, he indeed pays 10$ and not 9.782628$ ?

> 4. In any language other than English, customer order note is unreadable. This bug is open since Forge times. It is a matter of changing encoding/escaping somewhere in the code, I suppose changing like 3 lines of code. Just why?

Probably because it is not 3 lines of code, and probably because there are other bugs which are more critical which need our focus first.

> 5. Any customer in our store can go in to his Voucher screen in his account. He can see ALL VOUCHERS in the store, all vouchers with possible huge reductions that you "secretly" give to other customers. Do you understand how crucial mistake this is for the business? This is stuff that needs to be fixed in the next version, and not "it's on our roadmap".

This is not any customer, this bug has been [explored](https://github.com/PrestaShop/PrestaShop/issues/13002) and there is a very specific set of conditions that need to be true for this to happen (this set is met in your case), consequently it is a bug that rarely is visible and consequently it has not been considered an urgent issue. I'm sorry but so far you are the only known impacted user of this bug. We have to fix first the issues that impact everybody before fixing the issues that impact a single case.

> 6. I opened a ticket about layered navigation's page scrolling. Imagine you are a customer and you select a filter. What do you want to do next, see the products OR select another filter. So why THE FUCK the page scrolls intentionally to the top? Just WHY? And then @marionf tells me `No, it's not supposed to be fixed in 1.7.6.0. Please see this comment, the PR allow theme developer to choose to keep this scroll or not.` Show me one single person in the world that would want this behavior.

Again, same answer: we are not saying this is an intended behavior, we are saying there are more urgent matters that need to be done first, and when the urgent topics are done, we'll fix this one.

So, to sum it up: yes there are bugs, yes they need to be fixed and yes you think they are critical and should be fixed first. But we cannot only listen to you, we have to listen to *all* prestashop users, be it all kind of merchants, all kind of developers, all kind of agencies ... and we have to listen to all feedback and at the end define which are the most critical topics. We try to do our best but we cannot make everyone happy.

For example if I take your issues and make them my top priority issues for this week, you will be happy but that means the work that was assigned to me for this week will not be done, so someone else will be disappointed (because someone else is waiting for this work too). What do I tell this person ? What if I show him the issues you mentioned and he says that his issues are more important than yours ?

**Comment by eternoendless** [comment]
Hi @PrestaShark 

> Crowdfund is a keyword. Do it for crucial implementations. Do not hesitate to ask for it like Wikipedia do.

Wikipedia can because they are a nonprofit organization. But we are working on that 😉 

> Faceted search is most crucial right now guys...

Unfortunately everyone thinks _their_ issue is the most crucial one.

**Comment by Hlavtox** [comment]
@matks I am the only one that sees this bug? :D :D

Imagine a pretty normal business scenario.
- You have 3 default customer groups. You add 1 more for your long-term customers, maybe with better discounts or something like that.
- You make voucher with 20% reduction, which you occasionaly want to give somebody as a compensation for some trouble, or something like that.
- You restrict that voucher to the 3 default groups, so the 4. group cannot use it, if they found the code.
- Now, EVERY EVERY EVERY customer that makes account can see that 20% discount in his user account.

If you tell me again this is not a serious business-killing issue, than I came from Mars.

And how we found out? We randomly checked an invoice and some guy used all of our discounts that he could combine. I don't want to know how many people and other stores are loosing money because of this bug. This is serious guys!

**Comment by kpodemski** [comment]
> If PrestaShop cannot be used in a major country like Poland it would be a critical issue indeed. But as far as I know it is being used right now, so I dont think we can say "PS 1.7 is unusable in Poland".

> However I do not have a running Polish store, so I might be wrong on that. Consequently I will ask @kpodemski who is a renowned polish PrestaShop contributor whether this is a known issue and impacting issue. I think we can agree he's a good candidate to answer this question 😉

We have a lot successful projects on 1.7, but the fact is that dev costs has increased few times, just to have a budget to keep fighting with 1.7 ;-)

As for the "rounding" in PrestaShop, i have at least two customers who are patiently waiting for this to be available in PrestaShop, i thought that new CLDR feature would bring it back but unfortunately it's still not available.

For the translations, you know my opinion, i think that moving it to 1.7.7 for next 6 or more months is just bad, of course there are workarounds but if someone is trying to create multi-lang store, without developer, it's just a "no go" for him.

Issue with vouchers was reported by me almost a year ago:
https://github.com/PrestaShop/PrestaShop/issues/10766

of course i can make a PR but it would be moved to 1.7.7 for sure, so i decided to sit and wait if someone is even still aware of this issue, sorry guys :) I think that it's an important one, you think that it's not a regression  it's like: hey, we have this bug from 1.7 start! don't worry, it can wait! - really, when i see "it's not a regression, it can wait" i'm so angry 👎 

And i completely understand @Hlavtox - because situation is frustrating for me too, for example after 2 years from the release of PrestaShop 1.7, we still cannot even update from 1.6 to 1.7 properly without major issues like this one:
https://github.com/PrestaShop/PrestaShop/pull/14481

i said it many times - migration to Symfony should be done separately, i was fighting with the old team over 2 years ago asking them "why in the hell are you going to migrate only 2 pages to Symfony with so many other, major issues, creating spaghetti code here and there".

Current situation of PrestaShop is that PrestaShop 1.7 is transitional version and it's very hard for people, we have new pages migrated to Symfony, migration seems to be the best part of current development process and most important one. New release conditions are bad for people who want to have their stores stable, even tho merchant could sit and create an issue here on GitHub, waiting over 6 months for a fix is like a bad joke to him.

And you know what i'm afraid of? You will finish migration to Symfony and tell people "hey, we're creating PrestaShop v8! new and better! we will not support 1.7 in terms of new features and bugs, just major ones like spam etc." - and with this we will have customers so angry, that they new stores will based on e-commerce of magenta color. This is the future of PrestaShop 1.7 i see. Dark and sad, unfortunately.

**Comment by Hlavtox** [comment]
@eternoendless 

> Unfortunately everyone thinks their issue is the most crucial one.

I don't think my issues are the most crucial one, but what else do you want me to tell you?
You develop ECOMMERCE system without working currency management and proper rounding? Is there anything more crucial than money??? You are not making a CMS!

This software is a living for thousands of people, and what do you think just happened for example with this bug #15103? This made browsing thousands of store unsuable because the filter is just not working, and people are loosing money.

There are so many major issues that these issues would need another classification - Major as fuck, Major, Slightly less major and then the rest.

**Comment by ttoine** [comment]
Hello everyone involved in this discussion.

A few facts:

- 1.7.6 version is by far the most important version released, with the number of pull requests, of bug fixes, improvement and features delivered. The aim is to keep this productivity for next versions, and if possible to it. (1.7.4 was around 200 PR merged, 1.7.6 is around 800 PR merged)
- The work on important native module has started, the Facetized Search is a good example. Instead of complaining, maybe it's time also to rejoice, encourage developers and contributors, and thanks them for their work?
- The PrestaShop project has been open source from the start, but, managed more like a product than a project. This is going to change, the company is working on that, and the first step will be to add the maintainer role (people who can merge pull requests) to people who are not employees of the PrestaShop company. This will be a major change in the governance of the project. This is a work in progress, stay tuned for that.
- 1.7.x is not only a transition version, it is also mature enough to be used in production by more and more shops around the world. But, yes, for some people, it's complicated to upgrade, like with any software, being open source or not.
- currency and rounding issues are known issues, they are (very) old issues, and people are using PrestaShop anyway, so, the value of the software is still good enough. 1.7.6 solved parts of those issues, and 1.7.7 will continue to solve them.
- It's impossible to speak yet about a 1.8 or whatever will be its name. Let's finish the work on 1.7.x. And then, let's discuss about what's next. If you think about it, PrestaShop is the only CMS migrating to Symfony while supporting legacy (ask Drupal 7 users who want to move Drupal 8, and same for Magento 1/2 if they are happy). Yes, this is not perfect, but, this way of doing the migration actually let a lot of time to the ecosystem to learn the new technology stack, migrate when ready. The PrestaShop project is currently not following the "move fast break things" movement, it will not start with 1.7 and the following version.

If you feel that you need to discuss more about the direction of the project, [please open a dedicated issue](https://github.com/PrestaShop/open-source/issues), ping me in a comment, and add a link to it here. I will enjoy to discuss with you and answer your questions, take notes of your ideas, and see with you how to engage more contributors (we need more, we welcome more, and, we are working hard to be able to process more and more issues and pull requests). Let's use GitHub to speak about the PrestaShop open source project and its future.

That said and clarified, could we please come back to the discussion about this very issue?

**Comment by JustasBelevicius** [comment]
Any progress on this issue?

**Comment by PierreRambaud** [comment]
We are still waiting for QA on the faceted search pull requests. Be patient, it's coming :sweat_smile:

**Comment by ghost** [comment]
Unfortunately 3.2.1 version did not fix this issue but... keep up a good work @PierreRambaud !

**Comment by marionf** [comment]
@PrestaShark @Hlavtox @JustasBelevicius 

Could you please test this PR: https://github.com/PrestaShop/ps_facetedsearch/pull/132 and tell us if it solve the issue for you ?

**Comment by ghost** [comment]
With pleasure guys! Give me a minute!

edit.
1) On "shoe" store without change (something may be messed with DB)
2) Need to test on fresh enviroment 1.7.6.1

**Comment by Hlavtox** [comment]
@PrestaShark I don't use combinations on 1.7 yet, but this seems like a correct behavior.

What do you think is wrong?

- You select attribute Size 38,5. 
- It shows you all products, that have a combination with 38,5 attribute created.
- If you select "In Stock" filter, it doesn't look at quantity at all it seems. It just check if BO is enabled or disabled for the product.

Am I right @PierreRambaud ?

Proper stock filter has to be implemented yet - https://github.com/PrestaShop/PrestaShop/issues/14518#issuecomment-518119334

**Comment by ghost** [comment]
"In stock" filter should not even exist IMHO. That impact on UI/UX a lot...

By default In stock should be locked always as TRUE. Why customer might want to search by unavailable products?

Edit.
So when "In stock" filter is OFF then module should always show available products.

**Comment by Hlavtox** [comment]
@PrestaShark In this condition, yes. But a filter by quantity is needed A LOT.

**Comment by ghost** [comment]
@marionf @PierreRambaud additional feature.

When "In stock" filter is OFF (like in my shoe store) then module should always show available products.

**Comment by marionf** [comment]
@PrestaShark 

> When "In stock" filter is OFF (like in my shoe store) then module should always show available products.

_Available products_ = products in stock (quantity > 0) AND products out of stock (quantity <= 0) but available for order, right ?

**Comment by ghost** [comment]
@marionf 

Perfection ❤

**Comment by marionf** [comment]
Thank you for your feedback :)

So let's do it:
When "In stock" filter is OFF then show products in stock AND products out of stock but available for order

**Comment by Hlavtox** [comment]
When will you show products that are nit available for order then? Nonsense

**Comment by ghost** [comment]
Then 🤔
1) Remove "In stock" condition in faceted template

2) Make global option
"Search also unavailable products" YES / NO with default set as NO

**Comment by Hlavtox** [comment]
@marionf made the specification correct for all merchants.

All products
Availble products = only quantity > 0 or backorder on
In stock only = quantity > 0

**Comment by PierreRambaud** [comment]
It will be done in another pull request / issue. Better to don't mix issues in the same pull request :roll_eyes:

**Comment by ghost** [comment]
@PierreRambaud can You extract my idea from here?
Thank You for your time and effort. I own You a good beer and pizza!

@Hlavtox Your right. Need a good spec's for that. 

For now i can say that https://github.com/PrestaShop/ps_facetedsearch/pull/132 fix the mentioned issue

**Comment by marionf** [comment]
@PrestaShark please follow this issue: https://github.com/PrestaShop/PrestaShop/issues/14518#issuecomment-528247235

**Comment by matks** [comment]
🎉

**Comment by PierreRambaud** [comment]
Hey everyone, I updated the PrestaShop/ps_facetedsearch#132 pull request. Could you tell me if everything is ok for you?

**Comment by kossrlive** [comment]
Greetings) So that issue was solved?  
in ver 3.5 i can add filter by stock > 0. BUT i want only part of its realisation. 
I dont need that filter visible. But all my products must need to be filtered ONLY with available "in-stock" attributes. Its sounds like "Always in-stock mode". Even this filter not enabled.
Wish you all good guys

**Comment by MiroslavVo** [comment]
@kossrlive No, I think this issue was never solved, unfortunately.
